Signal/src/ViewControllers/ImagePickerController.swift

```diff
@@ -0,0 +1,439 @@
+            }
+        }
+    }
+
+    private func requestVideoDataSource(for asset: PHAsset) -> Promise<(dataSource: DataSource, dataUTI: String)> {
+        return Promise { resolver in
+
+            _ = imageManager.requestExportSession(forVideo: asset, options: nil, exportPreset: AVAssetExportPresetMediumQuality) { exportSession, info in
+
+                guard let exportSession = exportSession else {
+                    resolver.reject(PhotoLibraryError.assertionError("exportSession was unexpectedly nil"))
+                    return
+                }
+
+                exportSession.outputFileType = AVFileType.mp4
+                exportSession.metadataItemFilter = AVMetadataItemFilter.forSharing()
+
+                let exportPath = OWSFileSystem.temporaryFilePath(withFileExtension: "mp4")
+                let exportURL = URL(fileURLWithPath: exportPath)
+                exportSession.outputURL = exportURL
+
+                Logger.debug("starting video export")
+                exportSession.exportAsynchronously {
+                    Logger.debug("Completed video export")
+
+                    guard let dataSource = DataSourcePath.dataSource(with: exportURL, shouldDeleteOnDeallocation: true) else {
+                        resolver.reject(PhotoLibraryError.assertionError("Failed to build data source for exported video URL"))
+                        return
+                    }
+
+                    resolver.fulfill((dataSource: dataSource, dataUTI: kUTTypeMPEG4 as String))
+                }
+            }
+        }
+    }
+
+    func outgoingAttachment(for asset: PHAsset) -> Promise<SignalAttachment> {
+        switch asset.mediaType {
+        case .image:
+            return requestImageDataSource(for: asset).map { (dataSource: DataSource, dataUTI: String) in
+                return SignalAttachment.attachment(dataSource: dataSource, dataUTI: dataUTI, imageQuality: .medium)
+            }
+        case .video:
+            return requestVideoDataSource(for: asset).map { (dataSource: DataSource, dataUTI: String) in
+                return SignalAttachment.attachment(dataSource: dataSource, dataUTI: dataUTI)
+            }
+        default:
+            return Promise(error: PhotoLibraryError.unsupportedMediaType)
+        }
+    }
+}
+
+class PhotoLibrary: NSObject, PHPhotoLibraryChangeObserver {
+    weak var delegate: PhotoLibraryDelegate?
+
+    var assetCollection: PHAssetCollection!
+    var availableWidth: CGFloat = 0
+
+    func photoLibraryDidChange(_ changeInstance: PHChange) {
+        DispatchQueue.main.async {
+            self.delegate?.photoLibraryDidChange(self)
+        }
+    }
+
+    override init() {
+        super.init()
+        PHPhotoLibrary.shared().register(self)
+    }
+
+    deinit {
+        PHPhotoLibrary.shared().unregisterChangeObserver(self)
+    }
+
+    func albumForAllPhotos() -> PhotoLibraryAlbum {
+        let allPhotosOptions = PHFetchOptions()
```

Comment:
I suspect we'll want to only load a "window" into the album and raise the load window size to reflect scroll state. Otherwise, won't we be loading all assets into memory for users with large libraries?

Reply:
It does look reckless, but my understanding is that this is the correct way to load even very large collections, and that the Photo's framework does the work to make sure this performs well:

[excerpt from Photos framework](https://developer.apple.com/documentation/photokit/phfetchresult)

> You access the contents of a fetch result with the same methods and conventions used by the NSArray class. Unlike an NSArray object, however, a PHFetchResult object dynamically loads its contents from the Photos library as needed, providing optimal performance even when handling a large number of results.

Reply:
Furthermore, from what I can tell there is no way to do a "load window" or pagination even if we wanted to.